### ISSUE TITLE
Undefined Reference When Compiling `simple_xy_wr.cpp`

### ISSUE THREAD

**Issue by jasonlturner** [Unidata/netcdf-cxx4#99]
netCDF Version: 4.7.3

Environmental Information: I am encountering the same issue on both Windows 10 Education and Ubuntu 20.04.1 via WSL. I will describe how I produced the issue for both systems independently.

_Windows 10 Education_: I first asked for help on [StackOverflow about a week ago to no avail.](https://stackoverflow.com/questions/63940803/unresolved-external-symbols-when-using-netcdf-installed-via-vcpkg-with-cmake-i) For the sake of brevity here, I'll summarize: My coding environment on Windows 10 Education is Visual Studio. I installed `netcdf-cxx4` via `vcpkg`, and am using CMake to build the program I am linking to. In the project, which I named `cmake_netcdf` because that's what I am learning how to use, building the project gives the errors
```
>------ Build All started: Project: cmake_netcdf, Configuration: x64-Debug ------
  [1/2] Building CXX object CMakeFiles\cmake_netcdf.dir\main.cpp.obj
  [2/2] Linking CXX executable cmake_netcdf.exe
  FAILED: cmake_netcdf.exe 
  cmd.exe /C "cd . && "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E vs_link_exe --intdir=CMakeFiles\cmake_netcdf.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~1\2019\COMMUN~1\VC\Tools\MSVC\1427~1.291\bin\Hostx64\x64\link.exe /nologo CMakeFiles\cmake_netcdf.dir\main.cpp.obj  /out:cmake_netcdf.exe /implib:cmake_netcdf.lib /pdb:cmake_netcdf.pdb /version:0.0  /machine:x64 /debug /INCREMENTAL /subsystem:console  "C:\Program Files\vcpkg\installed\x64-windows\debug\lib\netcdf.lib"  "C:\Program Files\vcpkg\installed\x64-windows\debug\lib\hdf5_hl_D.lib"  "C:\Program Files\vcpkg\installed\x64-windows\debug\lib\hdf5_D.lib"  "C:\Program Files\vcpkg\installed\x64-windows\debug\lib\libcurl-d.lib"  "C:\Program Files\vcpkg\installed\x64-windows\debug\lib\zlibd.lib"  wldap32.lib  winmm.lib  ws2_32.lib  advapi32.lib  crypt32.lib  advapi32.lib  crypt32.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cmd.exe /C "cd /D "C:\Users\Owner\Education and Research\Personal Projects\learning_coding\C++\cmake_netcdf\build\x64-Debug" && powershell -noprofile -executionpolicy Bypass -file "C:/Program Files/vcpkg/scripts/buildsystems/msbuild/applocal.ps1" -targetBinary "C:/Users/Owner/Education and Research/Personal Projects/learning_coding/C++/cmake_netcdf/build/x64-Debug/cmake_netcdf.exe" -installedDir "C:/Program Files/vcpkg/installed/x64-windows/debug/bin" -OutVariable out""
  LINK Pass 1: command "C:\PROGRA~2\MICROS~1\2019\COMMUN~1\VC\Tools\MSVC\1427~1.291\bin\Hostx64\x64\link.exe /nologo CMakeFiles\cmake_netcdf.dir\main.cpp.obj /out:cmake_netcdf.exe /implib:cmake_netcdf.lib /pdb:cmake_netcdf.pdb /version:0.0 /machine:x64 /debug /INCREMENTAL /subsystem:console C:\Program Files\vcpkg\installed\x64-windows\debug\lib\netcdf.lib C:\Program Files\vcpkg\installed\x64-windows\debug\lib\hdf5_hl_D.lib C:\Program Files\vcpkg\installed\x64-windows\debug\lib\hdf5_D.lib C:\Program Files\vcpkg\installed\x64-windows\debug\lib\libcurl-d.lib C:\Program Files\vcpkg\installed\x64-windows\debug\lib\zlibd.lib wldap32.lib winmm.lib ws2_32.lib advapi32.lib crypt32.lib advapi32.lib crypt32.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:CMakeFiles\cmake_netcdf.dir/intermediate.manifest CMakeFiles\cmake_netcdf.dir/manifest.res" failed (exit code 1120) with the following output:
C:\Users\Owner\Education and Research\Personal Projects\learning_coding\C++\cmake_netcdf\build\x64-Debug\main.cpp.obj : error LNK2019: unresolved external symbol "public: __cdecl netCDF::NcDim::NcDim(class netCDF::NcDim const &)" (??0NcDim@netCDF@@QEAA@AEBV01@@Z) referenced in function "public: static void __cdecl std::_Default_allocator_traits<class std::allocator<class netCDF::NcDim> >::construct<class netCDF::NcDim,class netCDF::NcDim &>(class std::allocator<class netCDF::NcDim> &,class netCDF::NcDim * const,class netCDF::NcDim &)" (??$construct@VNcDim@netCDF@@AEAV12@@?$_Default_allocator_traits@V?$allocator@VNcDim@netCDF@@@std@@@std@@SAXAEAV?$allocator@VNcDim@netCDF@@@1@QEAVNcDim@netCDF@@AEAV34@@Z)
C:\Users\Owner\Education and Research\Personal Projects\learning_coding\C++\cmake_netcdf\build\x64-Debug\main.cpp.obj : error LNK2019: unresolved external symbol "public: class netCDF::NcVar __cdecl netCDF::NcGroup::addVar(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class netCDF::NcType const &,class std::vector<class netCDF::NcDim,class std::allocator<class netCDF::NcDim> > const &)const " (?addVar@NcGroup@netCDF@@QEBA?AVNcVar@2@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEBVNcType@2@AEBV?$vector@VNcDim@netCDF@@V?$allocator@VNcDim@netCDF@@@std@@@5@@Z) referenced in function main
C:\Users\Owner\Education and Research\Personal Projects\learning_coding\C++\cmake_netcdf\build\x64-Debug\main.cpp.obj : error LNK2019: unresolved external symbol "public: class netCDF::NcDim __cdecl netCDF::NcGroup::addDim(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,unsigned __int64)const " (?addDim@NcGroup@netCDF@@QEBA?AVNcDim@2@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@_K@Z) referenced in function main
C:\Users\Owner\Education and Research\Personal Projects\learning_coding\C++\cmake_netcdf\build\x64-Debug\main.cpp.obj : error LNK2019: unresolved external symbol "public: __cdecl netCDF::NcFile::NcFile(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,enum netCDF::NcFile::FileMode)" (??0NcFile@netCDF@@QEAA@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@W4FileMode@01@@Z) referenced in function main
C:\Users\Owner\Education and Research\Personal Projects\learning_coding\C++\cmake_netcdf\build\x64-Debug\main.cpp.obj : error LNK2019: unresolved external symbol "public: virtual __cdecl netCDF::NcFile::~NcFile(void)" (??1NcFile@netCDF@@UEAA@XZ) referenced in function main
C:\Users\Owner\Education and Research\Personal Projects\learning_coding\C++\cmake_netcdf\build\x64-Debug\main.cpp.obj : error LNK2019: unresolved external symbol "public: void __cdecl netCDF::NcVar::putVar(void const *)const " (?putVar@NcVar@netCDF@@QEBAXPEBX@Z) referenced in function main
C:\Users\Owner\Education and Research\Personal Projects\learning_coding\C++\cmake_netcdf\build\x64-Debug\main.cpp.obj : error LNK2001: unresolved external symbol "class netCDF::NcInt netCDF::ncInt" (?ncInt@netCDF@@3VNcInt@1@A)
C:\Users\Owner\Education and Research\Personal Projects\learning_coding\C++\cmake_netcdf\build\x64-Debug\cmake_netcdf.exe : fatal error LNK1120: 7 unresolved externals
  ninja: build stopped: subcommand failed.

Build All failed.
```

In the end, I am hoping to be able to build my project with CMake, so any help on that front would be especially appreciated!

_Ubuntu 20.40.1_: I installed `netCDF` via `apt-get` with the instructions supplied [here](https://cloud-gc.readthedocs.io/en/stable/chapter04_developer-guide/install-basic.html). I then installed `netcdf-cxx4` using the instructions included on this GitHub page. The command `nc-config --all` yields
```
This netCDF 4.7.3 has been built with the following features:

  --cc            -> /usr/bin/cc
  --cflags        -> -I/usr/include -I/usr/include/hdf5/serial
  --libs          -> -L/usr/lib/x86_64-linux-gnu -L/usr/lib/x86_64-linux-gnu/hdf5/serial -lnetcdf
  --static        -> -lhdf5_hl -lhdf5 -lpthread -lsz -lz -ldl -lm -lcurl

  --has-c++       -> no
  --cxx           ->

  --has-c++4      -> yes
  --cxx4          -> /usr/bin/c++
  --cxx4flags     -> -I/usr/local/include
  --cxx4libs      -> -L/usr/local/lib -lnetcdf-cxx4 -lnetcdf

  --has-fortran   -> yes
  --fc            -> gfortran
  --fflags        -> -I/usr/include
  --flibs         -> -L/usr/lib/x86_64-linux-gnu -lnetcdff -Wl,-Bsymbolic-functions -Wl,-z,relro -Wl,-z,now -lnetcdf -lnetcdf -ldl -lm
  --has-f90       ->
  --has-f03       -> yes

  --has-dap       -> yes
  --has-dap2      -> yes
  --has-dap4      -> yes
  --has-nc2       -> yes
  --has-nc4       -> yes
  --has-hdf5      -> yes
  --has-hdf4      -> no
  --has-logging   -> no
  --has-pnetcdf   -> no
  --has-szlib     -> no
  --has-cdf5      -> yes
  --has-parallel4 -> no
  --has-parallel  -> no

  --prefix        -> /usr
  --includedir    -> /usr/include
  --libdir        -> /usr/lib/x86_64-linux-gnu
  --version       -> netCDF 4.7.3
```
which appears to be in order. I am attempting to compile `SimpleXyWr.cpp`, whose contents are
```
/* This is part of the netCDF package.
   Copyright 2006 University Corporation for Atmospheric Research/Unidata.
   See COPYRIGHT file for conditions of use.

   This is a very simple example which writes a 2D array of
   sample data. To handle this in netCDF we create two shared
   dimensions, "x" and "y", and a netCDF variable, called "data".

   This example is part of the netCDF tutorial:
   http://www.unidata.ucar.edu/software/netcdf/docs/netcdf-tutorial

   Full documentation of the netCDF C++ API can be found at:
   http://www.unidata.ucar.edu/software/netcdf/docs/netcdf-cxx

   $Id: simple_xy_wr.cpp,v 1.5 2010/02/11 22:36:43 russ Exp $
*/

#include <iostream>
#include <netcdf>
#include <vector>
using namespace std;
using namespace netCDF;
using namespace netCDF::exceptions;

// We are writing 2D data, a 6 x 12 grid.
static const int NX = 6;
static const int NY = 12;

// Return this in event of a problem.
static const int NC_ERR = 2;



int main()
{
  // This is the data array we will write. It will just be filled
  // with a progression of numbers for this example.
  int dataOut[NX][NY];

  // Create some pretend data. If this wasn't an example program, we
  // would have some real data to write, for example, model output.
  for(int i = 0; i < NX; i++)
    for(int j = 0; j < NY; j++)
      dataOut[i][j] = i * NY + j;

  // The default behavior of the C++ API is to throw an exception i
  // an error occurs. A try catch block is necessary.

  try
    {
      // Create the file. The Replace parameter tells netCDF to overwrite
      // this file, if it already exists.
      NcFile dataFile("simple_xy.nc", NcFile::replace);

      // Create netCDF dimensions
      NcDim xDim = dataFile.addDim("x", NX);
      NcDim yDim = dataFile.addDim("y", NY);

      // Define the variable. The type of the variable in this case is
      // ncInt (32-bit integer).
      vector<NcDim> dims;
      dims.push_back(xDim);
      dims.push_back(yDim);
      NcVar data = dataFile.addVar("data", ncInt, dims);

      // Write the data to the file. Although netCDF supports
      // reading and writing subsets of data, in this case we write all
      // the data in one operation.
      data.putVar(dataOut);

      // The file will be automatically close when the NcFile object goes
      // out of scope. This frees up any internal netCDF resources
      // associated with the file, and flushes any buffers.

      //cout << "*** SUCCESS writing example file simple_xy.nc!" << endl;
      return 0;
    }
  catch(NcException& e)
    {e.what();
      return NC_ERR;
    }
}
```
The compile command I am using is `g++ -I/usr/local/include -L/usr/local/lib -lnetcdf-cxx4 -lnetcdf SimpleXyWr.cpp -o netcdf-example`, and it outputs
```
/usr/bin/ld: /tmp/ccqTGqi9.o: in function `main':
SimpleXyWr.cpp:(.text+0xe2): undefined reference to `netCDF::NcFile::NcFile(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, netCDF::NcFile::FileMode)'
/usr/bin/ld: SimpleXyWr.cpp:(.text+0x14e): undefined reference to `netCDF::NcGroup::addDim(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long) const'
/usr/bin/ld: SimpleXyWr.cpp:(.text+0x1ba): undefined reference to `netCDF::NcGroup::addDim(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long) const'
/usr/bin/ld: SimpleXyWr.cpp:(.text+0x26b): undefined reference to `netCDF::ncInt'
/usr/bin/ld: SimpleXyWr.cpp:(.text+0x273): undefined reference to `netCDF::NcGroup::addVar(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, netCDF::NcType const&, std::vector<netCDF::NcDim, std::allocator<netCDF::NcDim> > const&) const'
/usr/bin/ld: SimpleXyWr.cpp:(.text+0x2aa): undefined reference to `netCDF::NcVar::putVar(void const*) const'
/usr/bin/ld: SimpleXyWr.cpp:(.text+0x2fa): undefined reference to `netCDF::NcFile::~NcFile()'
/usr/bin/ld: SimpleXyWr.cpp:(.text+0x459): undefined reference to `netCDF::NcFile::~NcFile()'
/usr/bin/ld: /tmp/ccqTGqi9.o: in function `void __gnu_cxx::new_allocator<netCDF::NcDim>::construct<netCDF::NcDim, netCDF::NcDim const&>(netCDF::NcDim*, netCDF::NcDim const&)':
SimpleXyWr.cpp:(.text._ZN9__gnu_cxx13new_allocatorIN6netCDF5NcDimEE9constructIS2_JRKS2_EEEvPT_DpOT0_[_ZN9__gnu_cxx13new_allocatorIN6netCDF5NcDimEE9constructIS2_JRKS2_EEEvPT_DpOT0_]+0x47): undefined reference to `netCDF::NcDim::NcDim(netCDF::NcDim const&)'
/usr/bin/ld: /tmp/ccqTGqi9.o: in function `void std::_Construct<netCDF::NcDim, netCDF::NcDim const&>(netCDF::NcDim*, netCDF::NcDim const&)':
SimpleXyWr.cpp:(.text._ZSt10_ConstructIN6netCDF5NcDimEJRKS1_EEvPT_DpOT0_[_ZSt10_ConstructIN6netCDF5NcDimEJRKS1_EEvPT_DpOT0_]+0x43): undefined reference to `netCDF::NcDim::NcDim(netCDF::NcDim const&)'
/usr/bin/ld: /tmp/ccqTGqi9.o:(.data.rel.local.DW.ref._ZTIN6netCDF10exceptions11NcExceptionE[DW.ref._ZTIN6netCDF10exceptions11NcExceptionE]+0x0): undefined reference to `typeinfo for netCDF::exceptions::NcException'
collect2: error: ld returned 1 exit status
```

For extra information, the `include` and `lib` directories have the following files
```
- /usr/local/lib/
     - cmake/
          - netCDF/
               - netCDFCxxConfig.cmake
               - netCDFCxxConfigVersion.cmake
               - netcdf-cxx4Targets.debug.cmake
               - netcdf-cxx4Targets.cmake
     - libnetcdf-cxx.settings
     - libnetcdf-cxx4.so
     - libnetcdf-cxx4.so.1
     - libnetcdf-cxx4.so.1.1.0
     - python3.8/ (empty)
```
```
- usr/local/include/
     - ncAtt.h
     - ncChar.h
     - ncCompoundType.h
     - ncDouble.h
     - ncException.h
     - ncFill.h
     - ncFloat.h
     - ncGroupAtt.h
     - ncInt64.h
     - ncShort.h 
     - ncType.h
     - ncUint.h
     - ncUshort.h
     - ncVarAtt.h
     - netcdf
     - ncByte.h
     - ncCheck.h
     - ncDim.h
     - ncEnumType.h
     - ncFile.h
     - ncFilter.h
     - ncGroup.h
     - ncInt.h
     - ncOpaqueType.h
     - ncString.h
     - ncUbyte.h
     - ncUint64.h
     - ncVar.h
     - ncVlenType.h
     - test_utilities.h
```

Any ideas with there are these undefined references and what I can do about them?

**Comment by jasonlturner** [comment]
Okay, so the Ubuntu issue was easy enough to solve! I just changed the order of the compile flags:
```g++ SimpleXyWr.cpp -o netcdf-example -I/usr/local/include -L/usr/local/lib -lnetcdf-cxx4 -lnetcdf```

So my only question remaining is on how to link to `netCDF-cxx4` when using CMake. Since the Ubuntu issue was resolved so quickly, I will copy-paste the relevant information from the StackOverflow question I posted.

_Windows 10 Education_: In the project itself, I have two main files: `main.cpp` and `CMakeLists.txt`. The directory structure is as follows:
```
- C:\
     - Users\
          - Owner\
               - Education and Research\
                    - Personal Projects\
                         - learning_coding\
                              - C++\
                                   - cmake_netcdf\
                                        - CMakeLists.txt
                                        - main.cpp
```
The file 'main.cpp' is the same as `SimpleXyWr.cpp` from before, that is
```
/* This is part of the netCDF package.
   Copyright 2006 University Corporation for Atmospheric Research/Unidata.
   See COPYRIGHT file for conditions of use.

   This is a very simple example which writes a 2D array of
   sample data. To handle this in netCDF we create two shared
   dimensions, "x" and "y", and a netCDF variable, called "data".

   This example is part of the netCDF tutorial:
   http://www.unidata.ucar.edu/software/netcdf/docs/netcdf-tutorial

   Full documentation of the netCDF C++ API can be found at:
   http://www.unidata.ucar.edu/software/netcdf/docs/netcdf-cxx

   $Id: simple_xy_wr.cpp,v 1.5 2010/02/11 22:36:43 russ Exp $
*/

#include <iostream>
#include <netcdf>
#include <vector>
using namespace std;
using namespace netCDF;
using namespace netCDF::exceptions;

// We are writing 2D data, a 6 x 12 grid. 
static const int NX = 6;
static const int NY = 12;

// Return this in event of a problem.
static const int NC_ERR = 2;



int main()
{
    // This is the data array we will write. It will just be filled
    // with a progression of numbers for this example.
    int dataOut[NX][NY];

    // Create some pretend data. If this wasn't an example program, we
    // would have some real data to write, for example, model output.
    for (int i = 0; i < NX; i++)
        for (int j = 0; j < NY; j++)
            dataOut[i][j] = i * NY + j;

    // The default behavior of the C++ API is to throw an exception i
    // an error occurs. A try catch block is necessary.

    try
    {
        // Create the file. The Replace parameter tells netCDF to overwrite
        // this file, if it already exists.
        NcFile dataFile("simple_xy.nc", NcFile::replace);
        
        // Create netCDF dimensions
        NcDim xDim = dataFile.addDim("x", NX);
        NcDim yDim = dataFile.addDim("y", NY);
        
        // Define the variable. The type of the variable in this case is
        // ncInt (32-bit integer).
        vector<NcDim> dims;
        dims.push_back(xDim);
        dims.push_back(yDim);
        NcVar data = dataFile.addVar("data", ncInt, dims);

        // Write the data to the file. Although netCDF supports
        // reading and writing subsets of data, in this case we write all
        // the data in one operation.
        data.putVar(dataOut);

        // The file will be automatically close when the NcFile object goes
        // out of scope. This frees up any internal netCDF resources
        // associated with the file, and flushes any buffers.
        
        //cout << "*** SUCCESS writing example file simple_xy.nc!" << endl;
        return 0;
    }
    catch (NcException& e)
    {
        e.what();
        return NC_ERR;
    }

}
```
and the `CMakeLists.txt` file is
```
# Unit test for building a program with netCDF using CMake in visual studio.

cmake_minimum_required (VERSION 3.17)

project(cmake_netcdf)

# Add source to this project's executable.
add_executable(cmake_netcdf main.cpp)

# Find and include the netcdf package
find_package(netCDF CONFIG REQUIRED)
target_link_libraries(cmake_netcdf
                      PRIVATE netCDF::netcdf
                     )
```

When trying to build the project, I get the errors that I gave above, which are:
```
>------ Build All started: Project: cmake_netcdf, Configuration: x64-Debug ------
  [1/2] Building CXX object CMakeFiles\cmake_netcdf.dir\main.cpp.obj
  [2/2] Linking CXX executable cmake_netcdf.exe
  FAILED: cmake_netcdf.exe 
  cmd.exe /C "cd . && "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E vs_link_exe --intdir=CMakeFiles\cmake_netcdf.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~1\2019\COMMUN~1\VC\Tools\MSVC\1427~1.291\bin\Hostx64\x64\link.exe /nologo CMakeFiles\cmake_netcdf.dir\main.cpp.obj  /out:cmake_netcdf.exe /implib:cmake_netcdf.lib /pdb:cmake_netcdf.pdb /version:0.0  /machine:x64 /debug /INCREMENTAL /subsystem:console  "C:\Program Files\vcpkg\installed\x64-windows\debug\lib\netcdf.lib"  "C:\Program Files\vcpkg\installed\x64-windows\debug\lib\hdf5_hl_D.lib"  "C:\Program Files\vcpkg\installed\x64-windows\debug\lib\hdf5_D.lib"  "C:\Program Files\vcpkg\installed\x64-windows\debug\lib\libcurl-d.lib"  "C:\Program Files\vcpkg\installed\x64-windows\debug\lib\zlibd.lib"  wldap32.lib  winmm.lib  ws2_32.lib  advapi32.lib  crypt32.lib  advapi32.lib  crypt32.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cmd.exe /C "cd /D "C:\Users\Owner\Education and Research\Personal Projects\learning_coding\C++\cmake_netcdf\build\x64-Debug" && powershell -noprofile -executionpolicy Bypass -file "C:/Program Files/vcpkg/scripts/buildsystems/msbuild/applocal.ps1" -targetBinary "C:/Users/Owner/Education and Research/Personal Projects/learning_coding/C++/cmake_netcdf/build/x64-Debug/cmake_netcdf.exe" -installedDir "C:/Program Files/vcpkg/installed/x64-windows/debug/bin" -OutVariable out""
  LINK Pass 1: command "C:\PROGRA~2\MICROS~1\2019\COMMUN~1\VC\Tools\MSVC\1427~1.291\bin\Hostx64\x64\link.exe /nologo CMakeFiles\cmake_netcdf.dir\main.cpp.obj /out:cmake_netcdf.exe /implib:cmake_netcdf.lib /pdb:cmake_netcdf.pdb /version:0.0 /machine:x64 /debug /INCREMENTAL /subsystem:console C:\Program Files\vcpkg\installed\x64-windows\debug\lib\netcdf.lib C:\Program Files\vcpkg\installed\x64-windows\debug\lib\hdf5_hl_D.lib C:\Program Files\vcpkg\installed\x64-windows\debug\lib\hdf5_D.lib C:\Program Files\vcpkg\installed\x64-windows\debug\lib\libcurl-d.lib C:\Program Files\vcpkg\installed\x64-windows\debug\lib\zlibd.lib wldap32.lib winmm.lib ws2_32.lib advapi32.lib crypt32.lib advapi32.lib crypt32.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:CMakeFiles\cmake_netcdf.dir/intermediate.manifest CMakeFiles\cmake_netcdf.dir/manifest.res" failed (exit code 1120) with the following output:
C:\Users\Owner\Education and Research\Personal Projects\learning_coding\C++\cmake_netcdf\build\x64-Debug\main.cpp.obj : error LNK2019: unresolved external symbol "public: __cdecl netCDF::NcDim::NcDim(class netCDF::NcDim const &)" (??0NcDim@netCDF@@QEAA@AEBV01@@Z) referenced in function "public: static void __cdecl std::_Default_allocator_traits<class std::allocator<class netCDF::NcDim> >::construct<class netCDF::NcDim,class netCDF::NcDim &>(class std::allocator<class netCDF::NcDim> &,class netCDF::NcDim * const,class netCDF::NcDim &)" (??$construct@VNcDim@netCDF@@AEAV12@@?$_Default_allocator_traits@V?$allocator@VNcDim@netCDF@@@std@@@std@@SAXAEAV?$allocator@VNcDim@netCDF@@@1@QEAVNcDim@netCDF@@AEAV34@@Z)
C:\Users\Owner\Education and Research\Personal Projects\learning_coding\C++\cmake_netcdf\build\x64-Debug\main.cpp.obj : error LNK2019: unresolved external symbol "public: class netCDF::NcVar __cdecl netCDF::NcGroup::addVar(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class netCDF::NcType const &,class std::vector<class netCDF::NcDim,class std::allocator<class netCDF::NcDim> > const &)const " (?addVar@NcGroup@netCDF@@QEBA?AVNcVar@2@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEBVNcType@2@AEBV?$vector@VNcDim@netCDF@@V?$allocator@VNcDim@netCDF@@@std@@@5@@Z) referenced in function main
C:\Users\Owner\Education and Research\Personal Projects\learning_coding\C++\cmake_netcdf\build\x64-Debug\main.cpp.obj : error LNK2019: unresolved external symbol "public: class netCDF::NcDim __cdecl netCDF::NcGroup::addDim(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,unsigned __int64)const " (?addDim@NcGroup@netCDF@@QEBA?AVNcDim@2@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@_K@Z) referenced in function main
C:\Users\Owner\Education and Research\Personal Projects\learning_coding\C++\cmake_netcdf\build\x64-Debug\main.cpp.obj : error LNK2019: unresolved external symbol "public: __cdecl netCDF::NcFile::NcFile(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,enum netCDF::NcFile::FileMode)" (??0NcFile@netCDF@@QEAA@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@W4FileMode@01@@Z) referenced in function main
C:\Users\Owner\Education and Research\Personal Projects\learning_coding\C++\cmake_netcdf\build\x64-Debug\main.cpp.obj : error LNK2019: unresolved external symbol "public: virtual __cdecl netCDF::NcFile::~NcFile(void)" (??1NcFile@netCDF@@UEAA@XZ) referenced in function main
C:\Users\Owner\Education and Research\Personal Projects\learning_coding\C++\cmake_netcdf\build\x64-Debug\main.cpp.obj : error LNK2019: unresolved external symbol "public: void __cdecl netCDF::NcVar::putVar(void const *)const " (?putVar@NcVar@netCDF@@QEBAXPEBX@Z) referenced in function main
C:\Users\Owner\Education and Research\Personal Projects\learning_coding\C++\cmake_netcdf\build\x64-Debug\main.cpp.obj : error LNK2001: unresolved external symbol "class netCDF::NcInt netCDF::ncInt" (?ncInt@netCDF@@3VNcInt@1@A)
C:\Users\Owner\Education and Research\Personal Projects\learning_coding\C++\cmake_netcdf\build\x64-Debug\cmake_netcdf.exe : fatal error LNK1120: 7 unresolved externals
  ninja: build stopped: subcommand failed.

Build All failed.
```

I'm guessing there's another argument needed for `target_link_libraries`, but that doesn't seem to be `netcdf-cxx4`.

**Comment by ZedThree** [comment]
Have you tried the namespaced version, `netCDF::netcdf-cxx4`? This should be the target name, at least if you're using the current `master` branch

**Comment by jasonlturner** [comment]
I managed to find a solution for Ubuntu on WSL, which I posted about on the [StackOverflow](https://stackoverflow.com/questions/63940803/unresolved-external-symbols-when-using-netcdf-installed-via-vcpkg-with-cmake-i/64128682#64128682) post I had made. It essentially boiled down exactly to what you (@ZedThree) said, plus defining the path to the appropriate CMake file explicitly. I am unable to resolve the issue on Windows due to some issues with `vcpkg` with WSL enabled, but I no longer need that addressed.

Cheers!
